### ISSUE TITLE
RangeError: Maximum call stack size exceeded, When compiling

### ISSUE THREAD

**Issue by mx601595686** [microsoft/TypeScript#39138]
| package name | version |
|  ----  | ----  |
| typeScript|  3.9.5|
| gulp|  4.0.2|
| gulp-typescript|  5.0.1|
| eslint|  7.2.0|
| @typescript-eslint|  3.3.0|

**Search Terms:** RangeError

**This is my code on github:** [ObservableArray.ts](https://github.com/mx601595686/observable-variable/blob/master/src/ObservableDataType/ObservableArray.ts)

When I using `gulp-typescript` to compile it, or using `@typescript-eslint` to lint it, this error will occur.
And in vscode the syntax analysis will also be broken (Prompt a lot of error).
**This is `gulp-typescript` output:**
```
RangeError: Maximum call stack size exceeded
    at Object.assert (e:\observable-variable\node_modules\typescript\lib\typescript.js:2258:24)
    at lookupTypeParameterNodes (e:\observable-variable\node_modules\typescript\lib\typescript.js:39492:26)
    at createEntityNameFromSymbolChain (e:\observable-variable\node_modules\typescript\lib\typescript.js:39714:46)
    at symbolToName (e:\observable-variable\node_modules\typescript\lib\typescript.js:39712:24)
    at typeParameterToName (e:\observable-variable\node_modules\typescript\lib\typescript.js:39685:30)
    at typeParameterToDeclarationWithConstraint (e:\observable-variable\node_modules\typescript\lib\typescript.js:39337:28)
    at typeToTypeNodeHelper (e:\observable-variable\node_modules\typescript\lib\typescript.js:38771:55)
    at typeReferenceToTypeNode (e:\observable-variable\node_modules\typescript\lib\typescript.js:38968:43)
    at typeToTypeNodeHelper (e:\observable-variable\node_modules\typescript\lib\typescript.js:38766:95)
    at typeToTypeNodeHelper (e:\observable-variable\node_modules\typescript\lib\typescript.js:38823:43)
```

**This is `eslint` output:**
```
Oops! Something went wrong! :(

ESLint: 7.2.0

RangeError: Maximum call stack size exceeded
Occurred while linting e:\observable-variable\src\ObservableDataType\ObservableArray.ts:1
    at symbolToName (e:\observable-variable\node_modules\typescript\lib\typescript.js:39706:29)
    at typeParameterToName (e:\observable-variable\node_modules\typescript\lib\typescript.js:39685:30)
    at typeParameterToDeclarationWithConstraint (e:\observable-variable\node_modules\typescript\lib\typescript.js:39337:28)
    at typeToTypeNodeHelper (e:\observable-variable\node_modules\typescript\lib\typescript.js:38771:55)
    at typeReferenceToTypeNode (e:\observable-variable\node_modules\typescript\lib\typescript.js:38968:43)
    at typeToTypeNodeHelper (e:\observable-variable\node_modules\typescript\lib\typescript.js:38766:95)
    at typeToTypeNodeHelper (e:\observable-variable\node_modules\typescript\lib\typescript.js:38823:43)
    at mapToTypeNodes (e:\observable-variable\node_modules\typescript\lib\typescript.js:39236:40)
    at typeToTypeNodeHelper (e:\observable-variable\node_modules\typescript\lib\typescript.js:38790:37)
    at typeToTypeNodeHelper (e:\observable-variable\node_modules\typescript\lib\typescript.js:38825:40)
```

**Comment by ScarletFlash** [comment]
Same problem after upgrading from angular 9 (TS <3.9) to 10 (TS 3.9.5). – Now can not build project due to this issue.
My build output:

```bash
ERROR in RangeError: Maximum call stack size exceeded
    at typeParameterToName (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:39678:41)
    at typeParameterToDeclarationWithConstraint (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:39337:28)
    at typeToTypeNodeHelper (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:38771:55)
    at typeReferenceToTypeNode (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:38968:43)
    at typeToTypeNodeHelper (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:38766:95)
    at typeToTypeNodeHelper (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:38823:43)
    at mapToTypeNodes (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:39236:40)
    at typeToTypeNodeHelper (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:38790:37)
    at typeToTypeNodeHelper (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:38825:40)
    at mapToTypeNodes (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:39236:40)
    at typeToTypeNodeHelper (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:38790:37)
    at typeToTypeNodeHelper (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:38825:40)
    at mapToTypeNodes (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:39236:40)
    at typeToTypeNodeHelper (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:38790:37)
    at typeToTypeNodeHelper (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:38825:40)
    at mapToTypeNodes (/home/console/projects/___-__________/__________/___________frontend/node_modules/typescript/lib/typescript.js:39236:40)
```

**Comment by orta** [comment]
This has been fixed in 3.9.6, running the repo today gives an error and working inside doesn't seem to be crashing the TSServer:

```sh
❯ git clone https://github.com/mx601595686/observable-variable

❯ cd observable-variable

❯ cat node_modules/typescript/package.json | grep version
    "version": "3.9.6",

❯ yarn tsc
yarn run v1.22.4
$ /Users/ortatherox/dev/typescript/repros/39138/observable-variable/node_modules/.bin/tsc
src/ObservableDataType/ObservableArray.ts:445:9 - error TS2322: Type 'FlatArray<T, 0 | 1 | 2 | -1 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20>[]' is not assignable to type 'U[]'.
  Type 'FlatArray<T, 0 | 1 | 2 | -1 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20>' is not assignable to type 'U'.
    'U' could be instantiated with an arbitrary type which could be unrelated to 'FlatArray<T, 0 | 1 | 2 | -1 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20>'.
      Type 'T' is not assignable to type 'U'.
        'U' could be instantiated with an arbitrary type which could be unrelated to 'T'.

445         return this._value.flat(depth);
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error.

error Command failed with exit code 1.
```

vs

```sh
❯ yarn upgrade typescript@3.9.5
info Direct dependencies
└─ typescript@3.9.5
info All dependencies
└─ typescript@3.9.5
✨  Done in 9.43s.

❯ yarn tsc
yarn run v1.22.4
$ /Users/ortatherox/dev/typescript/repros/39138/observable-variable/node_modules/.bin/tsc
/Users/ortatherox/dev/typescript/repros/39138/observable-variable/node_modules/typescript/lib/tsc.js:82902
                throw e;
                ^

RangeError: Maximum call stack size exceeded
    at lookupTypeParameterNodes (/Users/ortatherox/dev/typescript/repros/39138/observable-variable/node_modules/typescript/lib/tsc.js:1:1)
    at createEntityNameFromSymbolChain (/Users/ortatherox/dev/typescript/repros/39138/observable-variable/node_modules/typescript/lib/tsc.js:32649:46)
    at symbolToName (/Users/ortatherox/dev/typescript/repros/39138/observable-variable/node_modules/typescript/lib/tsc.js:32647:24)
    at typeParameterToName (/Users/ortatherox/dev/typescript/repros/39138/observable-variable/node_modules/typescript/lib/tsc.js:32620:30)
    at typeParameterToDeclarationWithConstraint (/Users/ortatherox/dev/typescript/repros/39138/observable-variable/node_modules/typescript/lib/tsc.js:32300:28)
    at typeToTypeNodeHelper (/Users/ortatherox/dev/typescript/repros/39138/observable-variable/node_modules/typescript/lib/tsc.js:31769:55)
    at typeReferenceToTypeNode (/Users/ortatherox/dev/typescript/repros/39138/observable-variable/node_modules/typescript/lib/tsc.js:31954:43)
    at typeToTypeNodeHelper (/Users/ortatherox/dev/typescript/repros/39138/observable-variable/node_modules/typescript/lib/tsc.js:31764:95)
    at typeToTypeNodeHelper (/Users/ortatherox/dev/typescript/repros/39138/observable-variable/node_modules/typescript/lib/tsc.js:31819:43)
    at mapToTypeNodes (/Users/ortatherox/dev/typescript/repros/39138/observable-variable/node_modules/typescript/lib/tsc.js:32212:40)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
